### PR TITLE
Merge feature/create push notification to staging for Messaging 1.2.0 release

### PR DESCRIPTION
**Pull request by rymorale**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingPushNotificationFactory.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            notificationBuilder.setChannelId(channelId);
+        }
+        // Setting the default small icon
+        if (!setDefaultSmallIcon(context, notificationBuilder)) {
+            Log.warning(LOG_TAG, "%s - Failed to set the default small icon.", SELF_TAG);
+            return null;
+        }
+        // Setting the title
+        notificationBuilder.setContentTitle(payload.getTitle());
```
Comment (swarna04) [inline code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingPushNotificationFactory.java:73]: Shouldn't we verify that payload is not null?

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingPushNotificationFactory.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+     * @return a {@code boolean} containing true if the icon was set, false otherwise
+     */
+    private boolean setDefaultSmallIcon(final Context context, final Notification.Builder notificationBuilder) {
+        // Use the launcher icon as the default small icon.
+        int smallIcon = MessagingUtils.getDefaultSmallIconRes(context);
+        if (smallIcon == INVALID_SMALL_ICON_RES_ID) {
+            Log.warning(LOG_TAG, "%s - Invalid small icon found.", SELF_TAG);
+            return false;
+        }
+        notificationBuilder.setSmallIcon(smallIcon);
```
Comment (swarna04) [inline code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingPushNotificationFactory.java:141]: Verify non-null Notification builder is supplied.

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingUtils.java [hunk 3]
```diff
@@ -73,8 +93,310 @@ static List<Object> toList(final JSONArray jsonArray) throws JSONException {
...
+
+    /**
+     * Returns a {@link Uri} containing the location of a local audio file to use when displaying a push {@link Notification}.
+     *
+     * @param context  the application {@link Context}
+     * @param fileName the {@code String} containing the name of an audio file to use for a displayed notification
+     * @return {@code Uri} containing the local location of an audio file o use for a displayed {@code Notification}
+     */
+    static Uri getSoundUri(final Context context, final String fileName) {
+        final int resID = context.getResources().getIdentifier(fileName, "raw", context.getPackageName());
```
Comment (swarna04) [inline code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingUtils.java:186]: use camelCase for variable here.

**Review by swarna04 — CHANGES_REQUESTED** [review]

File: code/app/build.gradle [hunk 4]
```diff
@@ -40,13 +45,15 @@ def adobeDependencies = [
         edgeidentity   : "com.adobe.marketing.mobile:edgeidentity:+",
         lifecycle  : "com.adobe.marketing.mobile:lifecycle:+",
         signal     : "com.adobe.marketing.mobile:signal:+",
-        edge       : "com.adobe.marketing.mobile:edge:1.1.0"
+        edge       : "com.adobe.marketing.mobile:edge:1.1.0",
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/app/build.gradle:48]: This is the minimum required so I can use regex to get the latest 1.+. version.

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingInternal.java [hunk 5]
```diff
@@ -81,6 +80,142 @@ protected MessagingInternal(final ExtensionApi extensionApi) {
...
+
+    /**
+     * Adding XDM specific data to tracking information.
+     *
+     * @param eventData eventData which contains the xdm data forwarded by the customer.
+     * @param xdmMap    xdmMap map which is updated.
+     */
+    private static void addXDMData(final EventData eventData, final Map<String, Object> xdmMap) {
+        // Extract the xdm adobe data string from the event data.
+        final String adobe = eventData.optString(MessagingConstants.EventDataKeys.Messaging.TRACK_INFO_KEY_ADOBE_XDM, null);
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingInternal.java:162]: yes this is clearer

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                    singletonInstance = new MessagingDefaultImageDownloader();
+                }
+            }
+        }
+        return singletonInstance;
+    }
+
+    /**
+     * Downloads the asset then caches it in memory.
+     * TODO: Store the image asset in the disk cache added in the feature/iam branch
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java:70]: this is addressed in https://github.com/adobe/aepsdk-messaging-android/pull/88

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingImageDownloaderTask.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+     *
+     * @param url {@code String} containing the image asset url to be downloaded
+     * @return the downloaded image asset as a {@link Bitmap}
+     */
+    private Bitmap download(final String url) {
+        Bitmap bitmap = null;
+        HttpURLConnection connection = null;
+        try {
+            final URL imageUrl = new URL(url);
+            connection = (HttpURLConnection) imageUrl.openConnection();
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingImageDownloaderTask.java:61]: i've migrated to using the RemoteDownloader and our extension cache in this pr: https://github.com/adobe/aepsdk-messaging-android/pull/88

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        return bitmap;
+    }
+
+    /**
+     * Stores the provided {@link Bitmap} in the cache.
+     *
+     * @param key    The {@code String} key to use for caching the {@code Bitmap}
+     * @param bitmap a {@code Bitmap} to be cached
+     */
+    private void addBitmapToMemCache(String key, Bitmap bitmap) {
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java:111]: these functions are removed in https://github.com/adobe/aepsdk-messaging-android/pull/88

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    }
+
+    /**
+     * Retrieves a cached {@link Bitmap} from the cache.
+     *
+     * @param key The {@code String} key to use for retrieving the {@code Bitmap}
+     * @return {@code Bitmap} retrieved from the cache
+     */
+    private Bitmap getBitmapFromMemCache(String key) {
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java:123]: these functions are removed in https://github.com/adobe/aepsdk-messaging-android/pull/88

File: code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    private static volatile MessagingDefaultImageDownloader singletonInstance = null;
+    private final LruCache<String, Bitmap> cache;
+    private final ExecutorService executorService;
+
+    /**
+     * Constructor.
+     */
+    private MessagingDefaultImageDownloader() {
+        final int maxMemory = (int) (Runtime.getRuntime().maxMemory() / 1024);
+        final int cacheSize = maxMemory / 8;
```
[the remark this answers is not in the document]
Comment (rymorale) [reply code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingDefaultImageDownloader.java:39]: this is updated as well in https://github.com/adobe/aepsdk-messaging-android/pull/88

**Pull request closed without merging**